File: backend/src/OpenDriverHub.Infrastructure/Services/CategoryService.cs
using Microsoft.EntityFrameworkCore;
using OpenDriverHub.Application;
using OpenDriverHub.Domain;

namespace OpenDriverHub.Infrastructure.Services;

public class CategoryService : ICategoryService
{
    private readonly AppDbContext _db;
    public CategoryService(AppDbContext db) => _db = db;

    public async Task<List<CategoryDto>> ListAsync(CancellationToken ct)
        => (await _db.Categories
                .OrderBy(c => c.Type).ThenBy(c => c.Name).ToListAsync(ct))
            .Select(c => c.ToDto()).ToList();

    public async Task<CategoryDto> CreateAsync(CategoryUpsertRequest req, CancellationToken ct)
    {
        var name = req.Name.Trim();
        if (string.IsNullOrWhiteSpace(name))
            throw new AppException("Nome da categoria é obrigatório.", 400);
        var type = Mappings.ParseCategoryType(req.Type);
        if (await _db.Categories.AnyAsync(c => c.Name == name && c.Type == type, ct))
            throw new AppException("Já existe uma categoria com esse nome.", 409);
        var cat = new Category { Name = name, Type = type, Active = req.Active };
        _db.Categories.Add(cat);
        await _db.SaveChangesAsync(ct);
        return cat.ToDto();
    }

    public async Task<CategoryDto> UpdateAsync(Guid id, CategoryUpsertRequest req, CancellationToken ct)
    {
        var cat = await _db.Categories.FindAsync([id], ct)
            ?? throw new AppException("Categoria não encontrada.", 404);
        var name = req.Name.Trim();
        if (string.IsNullOrWhiteSpace(name))
            throw new AppException("Nome da categoria é obrigatório.", 400);
        if (await _db.Categories.AnyAsync(
                c => c.Name == name && c.Type == cat.Type && c.Id != id, ct))
            throw new AppException("Já existe uma categoria com esse nome.", 409);
        cat.Name = name;
        cat.Active = req.Active;
        await _db.SaveChangesAsync(ct);
        return cat.ToDto();
    }

    public async Task DeleteAsync(Guid id, CancellationToken ct)
    {
        var cat = await _db.Categories.FindAsync([id], ct)
            ?? throw new AppException("Categoria não encontrada.", 404);
        // Soft-delete: desativa para não quebrar produtos existentes.
        cat.Active = false;
        await _db.SaveChangesAsync(ct);
    }
}

File: backend/src/services/categoryService.ts
import type { CategoryDto, CategoryUpsertRequest } from '../dtos/catalog.dto.js';
import { AppError } from '../errors.js';
import { prisma } from '../infra/prisma.js';
import { parseCategoryType, toCategoryDto } from '../mappings.js';

export async function list(): Promise<CategoryDto[]> {
  const rows = await prisma.category.findMany({ orderBy: [{ type: 'asc' }, { name: 'asc' }] });
  return rows.map(toCategoryDto);
}

export async function create(req: CategoryUpsertRequest): Promise<CategoryDto> {
  const name = req.name.trim();
  if (!name) throw new AppError('Nome da categoria é obrigatório.', 400);
  const type = parseCategoryType(req.type);
  if (await prisma.category.findFirst({ where: { name, type } }))
    throw new AppError('Já existe uma categoria com esse nome.', 409);
  const cat = await prisma.category.create({ data: { name, type, active: req.active } });
  return toCategoryDto(cat);
}

export async function update(id: string, req: CategoryUpsertRequest): Promise<CategoryDto> {
  const cat = await prisma.category.findUnique({ where: { id } });
  if (!cat) throw new AppError('Categoria não encontrada.', 404);
  const name = req.name.trim();
  if (!name) throw new AppError('Nome da categoria é obrigatório.', 400);
  if (await prisma.category.findFirst({ where: { name, type: cat.type, id: { not: id } } }))
    throw new AppError('Já existe uma categoria com esse nome.', 409);

  const previousName = cat.name;
  const updated = await prisma.category.update({ where: { id }, data: { name, active: req.active } });
  if (previousName !== name) await propagateRename(cat.type, previousName, name);
  return toCategoryDto(updated);
}

export async function remove(id: string): Promise<void> {
  const cat = await prisma.category.findUnique({ where: { id } });
  if (!cat) throw new AppError('Categoria não encontrada.', 404);
  // Soft-delete: desativa para não quebrar produtos existentes.
  await prisma.category.update({ where: { id }, data: { active: false } });
}

async function propagateRename(type: 'Product' | 'Store', previousName: string, newName: string): Promise<void> {
  if (type === 'Product') {
    await prisma.product.updateMany({ where: { category: previousName }, data: { category: newName } });
    return;
  }
  await prisma.partner.updateMany({ where: { segment: previousName }, data: { segment: newName } });
  await prisma.partnerStore.updateMany({ where: { category: previousName }, data: { category: newName } });
}
